nucleus/common/common-util/src/main/java/com/sun/enterprise/util/io/FileUtils.java

```diff
@@ -551,8 +538,68 @@ else if (!deleteFile(f) && undeletedFiles != null) {
      * @param f file to delete
      */
     public static void deleteFileNowOrLater(File f) {
-        if(!deleteFile(f))
-            f.deleteOnExit();
+        if(!deleteFile(f)) {
+            deleteOnExit(f);
+        }
+    }
+
+    /**
+     * Deletes the provided file by registering a shutdown hook with the
+     * Java Runtime. Non-empty directories will be deleted recursively.
+     * @param f the file to delete on JVM shutdown
+     */
+    public static void deleteOnExit(File f) {
+        if (f != null && f.exists()) {
+            final SecurityManager securityManager = System.getSecurityManager();
+            if (securityManager != null) {
+                securityManager.checkDelete(f.getAbsolutePath());
+            }
+            FILE_DELETION_ON_EXIT.add(f);
+        }
+    }
+
+    private static final FileDeletionThread FILE_DELETION_ON_EXIT = new FileDeletionThread();
+
+    static {
+        Runtime.getRuntime().addShutdownHook(FILE_DELETION_ON_EXIT);
+    }
+
+    private static final class FileDeletionThread extends Thread {
+
+        private static final Logger LOG = Logger.getLogger(FileDeletionThread.class.getName());
+        private volatile Set<String> filesToDelete = new LinkedHashSet<>();
+
+        public void add(File file) {
+            if (file != null) {
+                filesToDelete.add(file.getAbsolutePath());
+            }
+        }
+
+        @Override
+        public void run() {
+            while (!filesToDelete.isEmpty()) {
+                final Set<String> filesToIterate = filesToDelete;
+                filesToDelete = new LinkedHashSet<>();
+                for (String path : filesToIterate) {
+                    delete(new File(path));
```

Comment:
Could `File`s (instead of `String`s) be stored at 570 and this construction skipped here?

Reply:
Sure. That was my first approach. I have changed it to String from File when @sgflt was concerend about the amount of memory this would consume and that this was a memory leak. Are you sure you want to go back to storing File instances?

Reply:
> Sure. That was my first approach...

I missed [that ...when @sgflt was concerned..](https://github.com/payara/Payara/pull/4201#discussion_r321582378) already about this case.

IMO keeping `File`s would be more clear.
I think you acted way too fast with https://github.com/payara/Payara/pull/4201/commits/9e8cccf2715f8b93460f43f57c60fbfe1c048cd7, as I'm 3rd party here :grin:

Reply:
We will see. I'm also for using Files.

Reply:
Files are OK for me. I was initially thinking about static array that is now gone. :) (moved to class as a field wich looks better to me)


```
final Set<String> filesToIterate = filesToDelete;
filesToDelete = new LinkedHashSet<>();
```
I was able to reproduce ConcurrentModificationException.
Without sync there is possible following sequence of steps:

T1 loads S1
T2 loads S1
T2 creates S2
T2 iterates over S1
T1 adds element to iterated S1